internals/api/encryption_key.go

```diff
@@ -0,0 +1,262 @@
+	KeyTypeAccountKey KeyType = "account-key"
+	KeyTypeSecretKey  KeyType = "secret-key"
+	KeyTypeAWS        KeyType = "aws"
+)
+
+// Options for KeyDerivationAlgorithm
+const (
+	KeyDerivationAlgorithmScrypt KeyDerivationAlgorithm = "scrypt"
+)
+
+// EncryptionKey specifies the common fields for all types of encryption keys.
+type EncryptionKey struct {
+	Type KeyType `json:"type"`
+}
+
+// NewEncryptionKeyDerivedScrypt creates a EncryptionKeyDerived with scrypt as key derivation algorithm.
+func NewEncryptionKeyDerivedScrypt(length, p, n, r int, salt []byte) *EncryptionKeyDerived {
+	return newEncryptionKeyDerived(
+		KeyDerivationAlgorithmScrypt,
+		length,
+		&KeyDerivationParametersScrypt{
+			P: Int(p),
+			N: Int(n),
+			R: Int(r),
+		},
+		&KeyDerivationMetadataScrypt{
+			Salt: salt,
+		},
+	)
+}
+
+func newEncryptionKeyDerived(algorithm KeyDerivationAlgorithm, length int, parameters, metadata interface{}) *EncryptionKeyDerived {
+	return &EncryptionKeyDerived{
+		EncryptionKey: EncryptionKey{
+			Type: KeyTypeDerived,
+		},
+		Length:     Int(length),
+		Algorithm:  algorithm,
+		Parameters: parameters,
+		Metadata:   metadata,
+	}
+}
+
+// EncryptionKeyDerived is an encryption key that can be derived from a passphrase.
+type EncryptionKeyDerived struct {
+	EncryptionKey
+	Length     *int                   `json:"length"`
+	Algorithm  KeyDerivationAlgorithm `json:"algorithm"`
+	Parameters interface{}            `json:"parameters,omitempty"`
+	Metadata   interface{}            `json:"metadata,omitempty"`
+}
+
+// AlgorithmSupported returns whether the given EncryptionAlgorithm is supported by this type of encrytpion key.
+func (EncryptionKeyDerived) AlgorithmSupported(a EncryptionAlgorithm) bool {
+	return a == EncryptionAlgorithmAESGCM
+}
+
+// Validate whether the EncryptionKeyDerived is valid.
+func (EncryptionKeyDerived) Validate() error {
+	// TODO: implement
+	return nil
+}
+
+// NewEncryptionKeyEncrypted creates a EncryptionKeyEncrypted.
+func NewEncryptionKeyEncrypted(length int, encryptedKey *EncryptedData) *EncryptionKeyEncrypted {
+	return &EncryptionKeyEncrypted{
+		EncryptionKey: EncryptionKey{
+			Type: KeyTypeEncrypted,
+		},
+		Length:       Int(length),
+		EncryptedKey: encryptedKey,
+	}
+}
+
+// EncryptionKeyEncrypted is an encryption key that has been encrypted by another key.
+type EncryptionKeyEncrypted struct {
+	EncryptionKey
+	Length       *int           `json:"length"`
+	EncryptedKey *EncryptedData `json:"encrypted_key"`
+}
+
+// AlgorithmSupported returns whether the given EncryptionAlgorithm is supported by this type of encrytpion key.
+func (EncryptionKeyEncrypted) AlgorithmSupported(a EncryptionAlgorithm) bool {
+	return a == EncryptionAlgorithmAESGCM || a == EncryptionAlgorithmRSAOEAP
+}
+
+// Validate whether the EncryptionKeyEncrypted is valid.
+func (k EncryptionKeyEncrypted) Validate() error {
+	if k.Length == nil {
+		return ErrMissingField("length")
+	}
+	if IntValue(k.Length) <= 0 {
```

Comment:
Doing some more sanity checks on used key lengths might be a good idea, yes. Though this one still is very thin if you did check for any of the proposed values because technically a user of the API can provide any value he likes here.